### ISSUE TITLE
LSTM for time series prediction

### ISSUE THREAD

**Issue by ashwinnaresh** [fchollet/keras#2856]
This is more of a conceptual clarification.

I have the following architecture : 

```
model = Sequential()
model.add(LSTM(batch_size,
               batch_input_shape=(batch_size, tsteps, 1)))  
model.compile(loss='mse', optimizer='rmsprop',metrics=["accuracy"])
```

The LSTM is able to make pretty accurate predictions as of now.

Training data : 
![issue_tr_le_full](https://cloud.githubusercontent.com/assets/5243800/15663143/15fc547a-2715-11e6-8aef-a8985f62d20c.png)

Note that the training data for each of the 3 weeks is almost similar in shape.

LSTM predictions on unseen testing data : 
![issue_test](https://cloud.githubusercontent.com/assets/5243800/15663157/521b2c56-2715-11e6-9cc6-b563d78ce9e9.png)

As you can see, the LSTM is still able to trace the unexpected peak in the graph which was never seen in the training data.

Any reason as to how the LSTM is able to do the above?

**Comment by gamehere007** [comment]
@ashwin123  
What is your input and output data for this model

**Comment by hijoe320** [comment]
@ashwin123 I am sure there is one step lag between the actual time series and the predicted time series, this is the most seen "trap" if you do time series prediction like this, in which the NN will always mimic previous input of time series.

**Comment by sjayakum** [comment]
@hijoe320  yup you are right, there was a 1 time-step lag between predicted and actual. the NN was mimicking the  previous input of the time series. 
Thanks!

**Comment by sjayakum** [comment]
@hijoe320  Ok, So are we doing any thing wrong to end up with this?

The main problem we have is the "Testing Data" real-time we get only 1 time-step at a time. So they are tested on a batch-size =  1 and time-step = 1.


Anyways to overcome this "trap"? :grin:  


-------------------------------------------

More information

```
batch_size = 1
tsteps = 1
```
We have trained **Each Timestep** as a sample.

Say lstm requires input of the form =>  nb_samples x tsteps x attributes

10:00 am -> X =  1*1*a  and y = b  goes into lstm    

10:01 am -> X = 1*1*b  and y = c goes into lstm.. 

10:02 am -> X = 1*1*c and y = d  goes into lstm.. And So on..

**Comment by hijoe320** [comment]
@sjayakum you can try:

1. randomize training samples in each batch, make sure they are not followed one by one
2. choose or design a better loss function other than MSE
3. extract some features from the input time series
4. manually limit the weight of x_{t-1}, x_{t-2}

**Comment by iwinterlovermy** [comment]
hi there. May I know how to use LSTM for multiple step ahead forecasting? Assuming I would like to predict 7 days ahead outcome. I've been trying few days but seems to get stuck - how should i prepare the data, and and what to set for batch size and time steps. Thanks in advance!

**Comment by safa94** [comment]
please did you find something to solve your problem i have the same problem here

**Comment by SunlnuS** [comment]
Hello, l'm using LSTM to predict time series and I meet the same problem. It seems that LSTM outputs input as output. Have you solved this problem? Would you like to give me some advice to solve this problem?

**Comment by shafferjohn** [comment]
In my opinion, loss function comes to minimum, so the network chooses to mimic the actual value. You will get same result even without LSTM, I think, even just with one fully connect layer.
